checks/process.go

```diff
@@ -106,28 +93,111 @@ func (p *ProcessCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.Mess
 	p.lastCPUTime = cpuTimes[0]
 	p.lastRun = time.Now()
 
-	statsd.Client.Gauge("datadog.process.containers.host_count", totalContainers, []string{}, 1)
-	statsd.Client.Gauge("datadog.process.processes.host_count", totalProcs, []string{}, 1)
+	statsd.Client.Gauge("datadog.process.containers.host_count", float64(totalContainers), []string{}, 1)
+	statsd.Client.Gauge("datadog.process.processes.host_count", float64(totalProcs), []string{}, 1)
 	log.Debugf("collected processes in %s", time.Now().Sub(start))
 	return messages, nil
 }
 
+func createProcCtrMessages(
+	procsByCtr map[string][]*model.Process,
+	containers []*model.Container,
+	cfg *config.AgentConfig,
+	sysInfo *model.SystemInfo,
+	groupID int32,
+) ([]model.MessageBody, int, int) {
+	totalProcs, totalContainers := 0, 0
+	msgs := make([]*model.CollectorProc, 0)
+
+	// we first split non-container processes in chunks
+	if procsByCtr[emptyCtrID] != nil {
+		totalProcs += len(procsByCtr[emptyCtrID])
+	}
+	chunks := chunkProcesses(procsByCtr[emptyCtrID], cfg.MaxPerMessage)
+	for _, c := range chunks {
+		msgs = append(msgs, &model.CollectorProc{
+			HostName:  cfg.HostName,
+			Info:      sysInfo,
+			Processes: c,
+			GroupId:   groupID,
+		})
+	}
+
+	// pack all containered processes with containers in a message
+	validIndex := 0
+	ctrProcs := make([]*model.Process, 0)
+	for _, ctr := range containers {
```

Comment:
Is there a way to move this filtering out of  `createProcCtrMessages` and pass the result in? It's sort of suprising that this function modifies `containers`

Reply:
I slightly disagree. the `containers` here is only used to create messages. Changing it in this function doesn't create any side effects. On the other hand, it's changed at the very end and isolates the logic well. But open to discussion just in case I'm wrong.

Reply:
`containers` is a parameter to this function and you're modifying it on line 139.  All I'm saying is that if someone tries to re-use `containers` after passing it to `createProcCtrMessages` they're going to be surprised.

Reply:
OK. I'll create a separate list inside the function and only append to it when needed.

Reply:
@leeavital this is addressed